BUG: move reduction initialization to ufunc initialization

ngoldbaum opened the pull request:
Fixes https://github.com/numpy/numpy/issues/28041.

Because the reduction initialization happens during ufunc initialization, this means that ufunc initialization now depends on the allocator context variables being initialized, so I moved those initialization steps up before ufunc initialization in the main `_multiarray_umath` module initialization function.

ngoldbaum committed:
BUG: move reduction initialization to ufunc initialization

ngoldbaum commented:
I'm also relying on the existing tests for reductions to cover this refactor - any suggestions for new tests?

seberg commented on numpy/_core/src/umath/legacy_array_method.c:
```diff
@@ -294,24 +294,7 @@ get_initial_from_ufunc(
-        }
-    }
-    else if (context->descriptors[0]->type_num == NPY_OBJECT
-            && !reduction_is_empty) {
```
Feels like moving this splits things even more maybe.

Maybe we can split this function to take `ufunc` and `descr` only (to avoid `context`, but we can also keep it this way and just abuse a `context`).

Then you can keep this here, and below only call the new function, something like:
```
if (PyTypeNum_ISNUMBER(bound_res->dtypes[0]->type_num)) {
    /* Cache our numerical reduce identity/initial to use `copy_cached_initial` */
    if (get_initial_from_ufunc(ufunc, bound_res->dtypes[0], bound_res->method->legacy_initial) < 0) {
        /* error */
    }
    bound_res->method->get_reduction_initial = &copy_cached_initial;
}
```
(It can already be simplified to write directly into the cache rather than via an allocation.)

About tests: I don't think new tests are needed, initials are well used and there is no subtlety here around default vs. initial.

ngoldbaum replied:
> I don't think new tests are needed, initials are well used and there is no subtlety here around default vs. initial.

Thanks for the confirmation and the code suggestions! I think I'll add the multithreaded test case in the issue report, that way this data race has a regression test when I add TSAN CI in a followup.

ngoldbaum committed:
MAINT: refactor to call get_initial_from_ufunc during init

ngoldbaum committed:
TST: add test for multithreaded reductions

ngoldbaum commented:
Refactored to call `get_initial_from_ufunc` during ufunc init, which indeed is a lot simpler and doesn't repeat as much code. There is a bit of extra boilerplate because I needed to heap-allocate a `descrs` array, but that's better than copy/pasting and is pretty straightforward.

Also added a test.

ngoldbaum committed:
MAINT: fix linter

seberg commented on numpy/_core/src/umath/legacy_array_method.c:
```diff
@@ -427,11 +420,67 @@ PyArray_NewLegacyWrappingArrayMethod(PyUFuncObject *ufunc,
+        }
+
+        PyArray_Descr **descrs = PyMem_Calloc(ufunc->nin + ufunc->nout,
+                                              sizeof(PyArray_Descr *));
```
But more importantly.  We only ever use the initial value for reductions.  So I think you might as well hard-code all of this (if you like up to the knowledge that `get_initial_form_ufunc()` only usees `descrs[0]` anyway).

So make `descrs[3]` or just `&singleton`...  (of course only enter the branch if `nin == 2` and `nout == 1`)

seberg commented on numpy/_core/src/umath/legacy_array_method.c:
```diff
@@ -427,11 +420,67 @@ PyArray_NewLegacyWrappingArrayMethod(PyUFuncObject *ufunc,
+
     Py_INCREF(res);
     Py_DECREF(bound_res);
+
```
`bound_res` isn't decref'ed in the errors above.  But once we remove all of the uneeded allocations, at least there is only one error path left.

seberg commented on numpy/_core/src/umath/legacy_array_method.c:
```diff
@@ -427,11 +420,67 @@ PyArray_NewLegacyWrappingArrayMethod(PyUFuncObject *ufunc,
+                descrs,
+        };
+
+        int ret = get_initial_from_ufunc(&context, 0, initial);
```
I think `PyArray_Pack` must support unaligned memory, so I think this is OK.

seberg reviewed (commented):
Thanks, nice to keep things together.  I still think it would be good to agressively simplify the new code, unless I am missing something?

Nice that you have a threading test!  (I am almost wondering if there is some pytest plugin that would run each tests N times in parallel, or if that could be a fun project... but, we certainly have a few bad test that might not survive that :), although a plugin could use a `pytest.mark` to run them only once)

ngoldbaum commented:
> I am almost wondering if there is some pytest plugin that would run each tests N times in parallel

There's [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) which SciPy is using heavily. Using it in NumPy may discover some things but I suspect it will mostly detect use of global state in tests and the signal-to-noise ratio may not be great. In particular the Python `warnings` module isn't thread-safe, so for SciPy the majority of the work ended up being skipping tests that use warnings filters - and there are a lot of tests for warnings!

For now I'm just adding one-off multithreaded tests for specific issues as I find them.

Thanks for the review!

ngoldbaum committed:
Apply suggestions from code review

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

ngoldbaum committed:
MAINT: simplify further

seberg reviewed (approved):
Thanks, LGTM to me now, let's put it in.

hawkinsp commented:
> > I am almost wondering if there is some pytest plugin that would run each tests N times in parallel
> 
> There's [pytest-run-parallel](https://github.com/Quansight-Labs/pytest-run-parallel) which SciPy is using heavily. Using it in NumPy may discover some things but I suspect it will mostly detect use of global state in tests and the signal-to-noise ratio may not be great. In particular the Python `warnings` module isn't thread-safe, so for SciPy the majority of the work ended up being skipping tests that use warnings filters - and there are a lot of tests for warnings!
> 
> For now I'm just adding one-off multithreaded tests for specific issues as I find them.
> 
> Thanks for the review!

There are ways to work around that (e.g. https://github.com/jax-ml/jax/pull/25626/files#diff-10c9477e1279e29626a5bbb8dccc529723495114f796523371741f472fc28e23) if you're sufficiently motivated. But it might be best just to wait for something like https://github.com/python/cpython/pull/128300 to land upstream.
